### PR TITLE
BUG: Let `melt` name multiple variable columns for labels from a `MultiIndex`

### LINKED ISSUE #58033: BUG: No longer can `melt` name the columns for the labels of a `MultiIndex`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

frame = pd.DataFrame({("A", "a"): range(3), ("A", "b"): range(3, 6)})
frame.melt(var_name=["upper", "lower"])
```

### Issue Description

Running the example gives
```Python traceback
Traceback (most recent call last):
  File "<string>", line 19, in __PYTHON_EL_eval
  File "/home/jdawson/Documents/bugs/pandas-melt/melt_bug.py", line 5, in <module>
    frame.melt(var_name=["upper", "lower"])
  File "/home/jdawson/Documents/bugs/pandas-melt/.venv/lib/python3.12/site-packages/pandas/core/frame.py", line 9839, in melt
    return melt(
           ^^^^^
  File "/home/jdawson/Documents/bugs/pandas-melt/.venv/lib/python3.12/site-packages/pandas/core/reshape/melt.py", line 220, in melt
    raise ValueError(f"{var_name=} must be a scalar.")
ValueError: var_name=['upper', 'lower'] must be a scalar.
```
We can melt the frame if we do not supply `var_name`, but the columns for the labels of the `MultiIndex` will have the abstruse names `variable_0` and `variable_1`.
```Python console
>>> frame.melt()
  variable_0 variable_1  value
0          A          a      0
1          A          a      1
2          A          a      2
3          A          b      3
4          A          b      4
5          A          b      5
```
This bug may be related to #57663.

### Expected Behavior

Versions strictly older than 2.2.0 behave thus.
```Python console
>>> import pandas as pd
>>> frame = pd.DataFrame({("A", "a"): range(3), ("A", "b"): range(3, 6)})
>>> frame
   A   
   a  b
0  0  3
1  1  4
2  2  5
>>> frame.melt(var_name=["upper", "lower"])
  upper lower  value
0     A     a      0
1     A     a      1
2     A     a      2
3     A     b      3
4     A     b      4
5     A     b      5
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : b552dc95c9fa50e9ca2a0c9f9cdb8757f794fedb
python                : 3.12.2.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 5.15.0-101-generic
Version               : #111~20.04.1-Ubuntu SMP Mon Mar 11 15:44:43 UTC 2024
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : en_GB.UTF-8
LOCALE                : en_GB.UTF-8

pandas                : 3.0.0.dev0+618.gb552dc95c
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : None
pip                   : 24.0
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by rob-sil**
- [x] closes #58033
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.2.2.rst` file if fixing a bug or adding a new feature.

If there are fewer names provided in `var_name` than levels in the columns `MultiIndex`, then `melt` will just take the first several. I personally find this a bit strange, but it keeps with the previous behavior of `melt` with a `MultiIndex` where you could pass it a single name and it would only use one level of the `MultiIndex`.

### CODE REVIEW

**Comment by Aloqeely** [comment]
Thanks for the PR! Sorry this hasn't been reviewed yet.
Generally LGTM. cc @mroeschke

**Comment by remi-pr** [comment]
Hi! Just checking in as this issue has been affecting my code. I wondered if anything needed to be done or if this will be merged in a later release?

File: pandas/tests/reshape/test_melt.py [hunk 1]
```diff
@@ -533,6 +533,22 @@ def test_melt_non_scalar_var_name_raises(self):
...
+        df = DataFrame({("A", "a"): [1], ("A", "b"): [2]})
+
+        expected = DataFrame(
+            [("A", "a", 1), ("A", "b", 2)], columns=["first", "second", "value"]
+        )
+
+        tm.assert_frame_equal(df.melt(var_name=["first", "second"]), expected)
+        tm.assert_frame_equal(df.melt(var_name=["first"]), expected[["first", "value"]])
+
+        with pytest.raises(
```
Comment (mroeschke) [inline pandas/tests/reshape/test_melt.py:547]: Could you make this a separate test?

File: pandas/core/reshape/melt.py [hunk 2]
```diff
@@ -217,7 +218,15 @@ def melt(
                 frame.columns.name if frame.columns.name is not None else "variable"
             ]
     elif is_list_like(var_name):
-        raise ValueError(f"{var_name=} must be a scalar.")
+        if isinstance(frame.columns, MultiIndex):
+            var_name = list(var_name)
```
Comment (mroeschke) [inline pandas/core/reshape/melt.py:222]: Is the `list` cast needed here? it should be `Sized` to compute `len`

File: doc/source/whatsnew/v2.2.2.rst [hunk 3]
```diff
@@ -15,6 +15,7 @@ Fixed regressions
 ~~~~~~~~~~~~~~~~~
 - :meth:`DataFrame.__dataframe__` was producing incorrect data buffers when the a column's type was a pandas nullable on with missing values (:issue:`56702`)
 - :meth:`DataFrame.__dataframe__` was producing incorrect data buffers when the a column's type was a pyarrow nullable on with missing values (:issue:`57664`)
+- :meth:`DataFrame.melt` would not accept multiple names in ``var_name`` when the columns were a :class:`MultiIndex` (:issue:`58033`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.2.2.rst:18]: Could you move this to `v3.0.0rst`?

File: pandas/core/reshape/melt.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/reshape/melt.py:222]: Ah true. Could you use `is_iterator` to check for the single use object and cast to `list` if so?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @rob-sil

**Comment by remi-pr** [comment]
Thanks for moving this forward!

**Comment by fuad47** [comment]
my code works in jupyter notebook wih 'var_names' parameter but not in py file, it gives the same error when running in py file. Interesting.

**Pull request merged**